src/absil/illib.fs

```diff
@@ -672,42 +672,44 @@ let _ = eventually { use x = null in return 1 }
-    let encodeTab = new Dictionary<'T,int>(HashIdentity.Structural)
-    let mutable nItems = 0
-    let encode str = 
-        if encodeTab.ContainsKey(str)
-        then
-            encodeTab.[str]
-        else
-            let idx = nItems
-            encodeTab.[str] <- idx
-            nItems <- nItems + 1
-            idx
-    member this.Encode(str)  = encode str
+[<Sealed>]
+type UniqueStampGenerator<'T when 'T : equality>(?initialCapacity) =
+    let encodeTab =
+        let initialCapacity = match initialCapacity with None -> 10 | Some c -> c
+        Dictionary<'T,int>(initialCapacity, HashIdentity.Structural)
+
+    member __.Encode(str) =
+        lock encodeTab <| fun () ->
+            let mutable strIdx = Unchecked.defaultof<_>
+            if encodeTab.TryGetValue (str, &strIdx) then strIdx
+            else
+                let idx = encodeTab.Count
+                encodeTab.Add (str, idx)
+                idx
 
 //---------------------------------------------------------------------------
 // memoize tables (all entries cached, never collected)
 //---------------------------------------------------------------------------
-    
-type MemoizationTable<'T,'U>(compute: 'T -> 'U, keyComparer: IEqualityComparer<'T>, ?canMemoize) = 
-    
-    let table = new System.Collections.Generic.Dictionary<'T,'U>(keyComparer) 
-    member t.Apply(x) = 
-        if (match canMemoize with None -> true | Some f -> f x) then 
-            let mutable res = Unchecked.defaultof<'U>
-            let ok = table.TryGetValue(x,&res)
-            if ok then res 
-            else
-                lock table (fun () -> 
-                    let mutable res = Unchecked.defaultof<'U> 
-                    let ok = table.TryGetValue(x,&res)
-                    if ok then res 
+[<Sealed>]
+type MemoizationTable<'T,'U>(compute: 'T -> 'U, keyComparer: IEqualityComparer<'T>, ?canMemoize, ?initialCapacity) =
+    let table =
+        let initialCapacity = match initialCapacity with None -> 100 | Some c -> c
+        System.Collections.Generic.Dictionary<'T, Lazy<'U>>(initialCapacity, keyComparer)
+    member __.Apply (x) =
+        if (match canMemoize with None -> true | Some f -> f x) then
+            let memoizedValue =
+                lock table <| fun () ->
+                    let mutable res = Unchecked.defaultof<Lazy<'U>>
+                    if table.TryGetValue (x, &res) then res
                     else
-                        let res = compute x
-                        table.[x] <- res;
-                        res)
+                        let lazyRes = lazy (compute x)
+                        table.Add (x, lazyRes)
+                        lazyRes
+
+            // Force evaluation of the computation for this value if it hasn't already been evaluated.
+            // Storing lazily-evaluated results in the table allows the computation to run outside
+            // of the table lock, so multiple results can be evaluated concurrently.
```

Comment:
For this reason, I don't think we should be adding fine-grained locking like this at this stage, since there's no intent that these particular objects will need to support high-performance fine-grained concurrent access in this way.  We can have a separate discussion about how to introduce more concurrency into the codebase, and where that might give good results (concurrent parsing may be a good start).